packages/jupyter-chat/src/components/input/use-chat-commands.tsx

```diff
@@ -0,0 +1,168 @@
+
+  // whether the chat commands menu is open
+  const [open, setOpen] = useState(true);
+
+  // current list of chat commands matched by the current word.
+  // the current word is the space-separated word at the user's cursor.
+  const [commands, setCommands] = useState<ChatCommand[]>([]);
+
+  useEffect(() => {
+    async function getCommands() {
+      const providers = chatCommandRegistry?.getProviders();
+      const cursorIndex = inputRef.current?.selectionStart;
+      if (!providers || cursorIndex === null || cursorIndex === undefined) {
+        return;
+      }
+
+      const currentWord = getCurrentWord(input, cursorIndex);
+      if (!currentWord?.length) {
+        return;
+      }
+
+      let newCommands: ChatCommand[] = [];
+      for (const provider of providers) {
+        // TODO: optimize performance when this method is truly async
+        try {
+          newCommands = newCommands.concat(
+            await provider.getChatCommands(currentWord)
+          );
+        } catch (e) {
+          console.error(
+            `Error when getting chat commands from command provider '${provider.id}': `,
+            e
+          );
+        }
+      }
+
+      if (newCommands) {
+        setOpen(true);
+      }
+      setCommands(newCommands);
+    }
+
+    getCommands();
+  }, [input, inputRef]);
+
+  return {
+    autocompleteProps: {
+      open,
+      options: commands,
+      getOptionLabel: (command: ChatCommand) => command.name,
+      renderOption: (
+        defaultProps,
+        command: ChatCommand,
+        __: unknown,
+        ___: unknown
+      ) => {
+        const { key, ...listItemProps } = defaultProps;
+        return (
+          <Box key={key} component="li" {...listItemProps}>
+            {command.name}
+          </Box>
+        );
+      },
+      value: null,
+      autoHighlight: true,
+      freeSolo: true,
+      disableClearable: true,
+      onChange: (
+        _: unknown,
+        command: ChatCommand,
+        reason: AutocompleteChangeReason
+      ) => {
+        if (reason !== 'selectOption') {
+          // only call this callback when a command is selected by the user. the
+          // other reasons provided by MUI should never occur, so this check is
+          // mainly for type safety.
+          return;
+        }
+
+        const cursorIndex = inputRef.current?.selectionStart;
+        if (
+          !chatCommandRegistry ||
+          cursorIndex === null ||
+          cursorIndex === undefined
+        ) {
+          return;
+        }
+
+        const currentWord = getCurrentWord(input, cursorIndex);
+        if (!currentWord) {
+          return;
+        }
+
+        const replaceCurrentWordClosure = (newWord: string) => {
+          replaceCurrentWord(input, cursorIndex, newWord, setInput);
+        };
+
+        chatCommandRegistry.handleChatCommand(
+          command,
+          currentWord,
+          replaceCurrentWordClosure
+        );
```

Comment:
Couldn't we replace the word in the `onChange()` callback directly. Since we have the command object, we could check if the attribute `replaceWith` is defined and replace the string. 

```ts
if (command.replaceWith) {
  replaceCurrentWord(input, cursorIndex, command.replaceWith, setInput);
}
```

Is there any advantage to replace it from the `handleChatCommand()` function ?

Reply:
Good thinking. I'll update the code & resolve this thread when I do so.

Reply:
Done!